### PR TITLE
Initializing respective weights by different method for TorchLayer

### PR DESCRIPTION
**Pull request by amitjansc**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

  **Note:** Running `make docs` raises the following error: `You must configure the bibtex_bibfiles setting`

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

The `init_method` argument of `TorchLayer`'s constructor has been re-defined. Now it can take two different types:
1. A `torch.nn.init` method to initialize all weights.
2. A dictionary containing a `torch.nn.init` method or a `torch.Tensor` for each different weight.

Depending on the `init_method` used, the newly defined `_init_weights` method initializes and registers all the weights.

This PR also solves a small bug, where the user could define weights with negative shape and these would be assigned to a scalar Tensor.

**Description of the Change:**

- `pennylane/qnn/torch.py`:
  - Moved weight initialization inside `_init_weights()`: 
    
      This method defines another method (`init_weight`) which returns the initialized weights depending on the given `init_method`. It then uses the aforementioned method to initialize all the weights defined in `weight_shapes`.
  - Changed `if` condition when building the `weight_shapes`, such that only shapes equal to 0 or 1 are casted to an empty tuple.
  - Style changes (a2ba68ee7571a6d66acbea8f49181757648953ef): Merge nested if conditions. Convert for loop into list comprehension. 
  - Added missing `QNode` type.

- `tests/qnn/test_qnn_torch.py`: Added tests with all possible types of init_method arguments.

**Benefits:**

In `TorchLayer`, we can now specify per weight initialization method. Consequently, it is possible to initialize, for example, multiple entangling layers in one `qnode` with different initialization methods.

**Possible Drawbacks:**

If `init_method` is a dictionary, it must contain the same keys than the `weight_shapes` dictionary. If not a `KeyError` is raised. I can add a small check to raise a more self-explanatory error if needed.

`torch.Tensor` cannot be used when defining the input/output types of the `TorchLayer` class methods, because it crashes the tests that don't use `torch`.

**Related GitHub Issues:**

Issue https://github.com/PennyLaneAI/pennylane/issues/898.

### CODE REVIEW

**Comment by antalszava** [comment]
Hi @amitjansc, thank you for this PR! :slightly_smiling_face: :tada: Let us know if we should go for a review. :+1:

**Comment by amitjansc** [comment]
Hi @antalszava,

I added a small commit checking the size of the given Tensor in the `init_method`. If all workflow checks are successful then the PR can be reviewed.

Please find below a working example where I initialize the weights using different methods and/or specific Tensor values:
```
import torch
import pennylane as qml

n_qubits = 1
dev = qml.device("default.qubit", wires=n_qubits)


@qml.qnode(dev)
def qnode(inputs, weights_0, weights_1, weights_2, weight_3, weight_4):
    qml.templates.AngleEmbedding(inputs, wires=range(n_qubits))
    qml.templates.StronglyEntanglingLayers(weights_0, wires=range(n_qubits))
    qml.templates.BasicEntanglerLayers(weights_1, wires=range(n_qubits))
    qml.Rot(*weights_2, wires=0)
    qml.RY(weight_3, wires=1)
    qml.RZ(weight_4, wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))


weight_shapes = {
    "weights_0": (3, n_qubits, 3),
    "weights_1": (3, n_qubits, 3),
    "weights_2": 3,
    "weight_3": 1,
    "weight_4": (1,),
}

init_method = {
    "weights_0": torch.nn.init.normal_,
    "weights_1": torch.nn.init.uniform,
    "weights_2": torch.tensor([1., 2., 3.]),
    "weight_3": torch.tensor(1.),  # scalar when shape is not an iterable and is <= 1
    "weight_4": torch.tensor([1.]),
}

qlayer = qml.qnn.TorchLayer(qnode, weight_shapes=weight_shapes, init_method=init_method)
print(qlayer.qnode_weights)
```
Prints:
```
{'weights_0': Parameter containing:
tensor([[[ 1.8947e-04, -5.3276e-01,  4.6640e-01]],

        [[-1.8541e+00, -6.5048e-01, -3.3519e-01]],

        [[ 3.0801e-01,  4.2132e-01,  1.0647e+00]]], requires_grad=True), 'weights_1': Parameter containing:
tensor([[[0.5746, 0.1815, 0.6344]],

        [[0.7437, 0.7722, 0.1158]],

        [[0.4531, 0.5425, 0.8562]]], requires_grad=True), 'weights_2': Parameter containing:
tensor([1., 2., 3.], requires_grad=True), 'weight_3': Parameter containing:
tensor(1., requires_grad=True), 'weight_4': Parameter containing:
tensor([1.], requires_grad=True)}
```

File: pennylane/qnn/torch.py [hunk 1]
```diff
@@ -291,6 +295,58 @@ def _evaluate_qnode(self, x):
...
+            elif isinstance(init_method, dict):
+                init = init_method[weight_name]
+                if isinstance(init, torch.Tensor):
+                    if tuple(init.shape) != weight_size:
+                        raise ValueError(
+                            f"The Tensor specified for weight '{weight_name}' doesn't have the"
+                            + "appropiate shape."
+                        )
+                    return init
+            # TODO: Figure out why a 1-dim Tensor is assigned to all weights with shape <= 1.
```
Comment (antalszava) [inline pennylane/qnn/torch.py:340]: Would this comment be out of date with [Tom's comment](https://github.com/PennyLaneAI/pennylane/issues/898#issuecomment-1151616249)?

File: pennylane/qnn/torch.py [hunk 2]
```diff
@@ -208,7 +219,9 @@ def __init__(self, qnode, weight_shapes: dict, init_method: Optional[Callable] =
         super().__init__()
 
         weight_shapes = {
-            weight: (tuple(size) if isinstance(size, Iterable) else (size,) if size > 1 else ())
+            weight: (
+                tuple(size) if isinstance(size, Iterable) else () if size in [0, 1] else (size,)
```
Comment (antalszava) [inline pennylane/qnn/torch.py:223]: Minor, as per Tom's comment. Can be ignored if preferred.
```suggestion
                tuple(size) if isinstance(size, Iterable) else () if size in [1] else (size,)
```

**Review by antalszava — APPROVED** [review]
Looks good @amitjansc! :tada: Thank you for the contribution. :slightly_smiling_face: 

Just a couple of minor comments, the addition looks great otherwise. :)

File: tests/qnn/test_qnn_torch.py [hunk 3]
```diff
@@ -173,6 +173,72 @@ def test_nonspecified_init(self, get_circuit, n_qubits, monkeypatch):
...
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
+    def test_fixed_and_callable_init(self, get_circuit, n_qubits):
+        """Test if weights are initialized according to the callables and values specified in the
+        init_method argument."""
+        c, w = get_circuit
+
+        normal_ = mock.MagicMock(return_value=torch.normal(mean=0, std=1, size=[1]))
+
+        init_method = {
+            "w1": normal_,
```
[the remark this answers is not in the document]
Comment (amitjansc) [reply tests/qnn/test_qnn_torch.py:240]: Done! As I mentioned above, when a weight is missing a KeyError is raised by Python. But I could implement a check assuring that weight_shapes and init_method have the same keys if needed.

File: pennylane/qnn/torch.py (hunk 1 above)
Comment (amitjansc) [reply pennylane/qnn/torch.py]: Yes, sorry. The comment has been removed.

File: pennylane/qnn/torch.py (hunk 2 above)
Comment (amitjansc) [reply pennylane/qnn/torch.py]: I changed this in commit https://github.com/PennyLaneAI/pennylane/pull/2678/commits/339ab96f7cbbef889361176c1c5aefe3a60c6415. It is important to know that this small change affects the usage of the `TorchLayer` class. Before a user could use `size = 0` or `size = 1` to define a scalar tensor. Now only `size = 1` can be used. Using `size = 0` will create an empty Tensor, which I don't think has any usage.

**Comment by amitjansc** [comment]
All comments have been addressed @antalszava :tada: :. 

Before merging, beware that I could not run `make docs` because I get the following error message (which I don't know how to solve :disappointed:) : `You must configure the bibtex_bibfiles setting`

File: pennylane/qnn/torch.py (hunk 1 above)
Comment (antalszava) [reply pennylane/qnn/torch.py]: No problem!

File: tests/qnn/test_qnn_torch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/qnn/test_qnn_torch.py:256]: Thanks :)

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -565,6 +565,9 @@
 * Control values are now displayed distinctly in text and mpl drawings of circuits.
   [(#2668)](https://github.com/PennyLaneAI/pennylane/pull/2668)
 
+* The `TorchLayer`'s `init_method` argument now accepts either a `torch.nn.init` function or a dictionary which should specify a `torch.nn.init`/`torch.Tensor` for each different weight.
+  [(#2678)](https://github.com/PennyLaneAI/pennylane/pull/2678)
+
 <h3>Breaking changes</h3>
```
Comment (antalszava) [inline doc/releases/changelog-dev.md:571]:
```suggestion
<h3>Breaking changes</h3>

* Specifying weights with negative shapes will raise an error, whereas specifying
  `size = 0` for weights will now create empty Tensor objects with `qml.TorchLayer`.
  [(#2678)](https://github.com/PennyLaneAI/pennylane/pull/2678)

```

File: pennylane/qnn/torch.py (hunk 2 above)
Comment (antalszava) [reply pennylane/qnn/torch.py]: Thank you! I've added a [suggested comment to the changelog](https://github.com/PennyLaneAI/pennylane/pull/2678/files#r894860328) to document this change in behaviour as a breaking change. The entry also mentioned the behaviour change for the definition of weights with negative shapes.

Comment (amitjansc) [reply pennylane/qnn/torch.py]: Everything good! 👍

**Comment by antalszava** [comment]
Hi @amitjansc, as you've made your contribution during Unitary HACK you are eligible for swag. :slightly_smiling_face: To claim it you should be signed up for UnitaryHACK [here](https://unitaryhack.dev/signup/) and add [unitaryhack] in the title of the PR.

Let me know if you have further questions (cc @CatalinaAlbornoz).

**Comment by amitjansc** [comment]
> Hi @amitjansc, as you've made your contribution during Unitary HACK you are eligible for swag. slightly_smiling_face To claim it you should be signed up for UnitaryHACK [here](https://unitaryhack.dev/signup/) and add [unitaryhack] in the title of the PR.
> 
> Let me know if you have further questions (cc @CatalinaAlbornoz).

Hi @antalszava. Great, thank you! I just signed up for UnitaryHACK.

**Pull request merged**